File: Sources/UsageWidget/Providers/Claude/ClaudeProvider.swift
import Foundation
import AppKit

/// Claude provider: paste-code OAuth sign-in, token refresh (proactive + reactive), and usage fetch.
actor ClaudeProvider: UsageProvider {
    nonisolated let id = ProviderID.claude
    nonisolated let displayName = "Claude"
    nonisolated let iconSystemName = "sparkle"
    nonisolated let accentHex = "#D97757"
    nonisolated let defaultPollInterval: TimeInterval = 300
    nonisolated let minimumPollInterval: TimeInterval = 300

    private let tokens: TokenStore
    private let oauth = ClaudeOAuthClient()
    private let fetcher = ClaudeUsageFetcher()
    private let profile = ClaudeProfileFetcher()

    init(tokens: TokenStore) {
        self.tokens = tokens
    }

    func authState() async -> ProviderAuthState {
        await tokens.token(for: id) != nil ? .signedIn : .signedOut
    }

    func startSignIn() async throws -> SignInContinuation {
        let pkce = PKCE.generate()
        let url = oauth.makeAuthorizeURL(pkce: pkce)
        await openInBrowser(url)

        let oauth = self.oauth
        let tokens = self.tokens
        let id = self.id
        let instructions = "Approve access in your browser, then paste the code it shows (looks like \"abc123#xyz\")."
        return .needsCode(instructions: instructions) { pastedCode in
            let token = try await oauth.exchange(pastedCode: pastedCode, pkce: pkce)
            await tokens.save(token, for: id)
        }
    }

    func signOut() async {
        await tokens.clear(id)
    }

    func fetchUsage() async throws -> ProviderUsage {
        guard var token = await tokens.token(for: id) else { throw ProviderError.notSignedIn }
        if token.needsRefresh() {
            token = try await refreshAndStore(token)
        }
        token = await ensurePlan(token)
        do {
            return try await fetcher.fetch(accessToken: token.accessToken, plan: token.planType)
        } catch ProviderError.unauthorized {
            var refreshed = try await refreshAndStore(token)
            refreshed = await ensurePlan(refreshed)
            return try await fetcher.fetch(accessToken: refreshed.accessToken, plan: refreshed.planType)
        }
    }

    /// The usage endpoint carries no plan, so fetch the profile once and cache it on the token.
    /// Best-effort: a failure just leaves the badge off — usage still loads.
    private func ensurePlan(_ token: OAuthToken) async -> OAuthToken {
        guard token.planType == nil else { return token }
        guard let raw = (try? await profile.fetchRawPlan(accessToken: token.accessToken)) ?? nil else {
            return token
        }
        var updated = token
        updated.planType = raw
        await tokens.save(updated, for: id)
        return updated
    }

    private func refreshAndStore(_ token: OAuthToken) async throws -> OAuthToken {
        do {
            var refreshed = try await oauth.refresh(token)
            refreshed.planType = refreshed.planType ?? token.planType   // keep the cached plan
            await tokens.save(refreshed, for: id)
            return refreshed
        } catch {
            await tokens.clear(id)
            throw ProviderError.unauthorized
        }
    }

    @MainActor
    private func openInBrowser(_ url: URL) {
        NSWorkspace.shared.open(url)
    }
}

File: windows/src/Core/Providers/ClaudeProvider.cs
using LLMUsageWidget.Core.Auth;
using LLMUsageWidget.Core.Domain;

namespace LLMUsageWidget.Core.Providers;

/// <summary>Claude provider: paste-code OAuth, token refresh, and usage fetch. The plan badge is
/// fetched once from the profile endpoint and cached on the token.</summary>
public sealed class ClaudeProvider : IUsageProvider
{
    public ProviderId Id => ProviderId.Claude;
    public string DisplayName => "Claude";
    public string AccentHex => "#D97757";
    public TimeSpan DefaultPollInterval => TimeSpan.FromSeconds(300);
    public TimeSpan MinimumPollInterval => TimeSpan.FromSeconds(300);

    private readonly ITokenStore _tokens;
    private readonly Action<Uri> _openBrowser;
    private readonly ClaudeOAuthClient _oauth = new();
    private readonly ClaudeUsageFetcher _fetcher = new();
    private readonly ClaudeProfileFetcher _profile = new();

    public ClaudeProvider(ITokenStore tokens, Action<Uri> openBrowser)
    {
        _tokens = tokens;
        _openBrowser = openBrowser;
    }

    public Task<ProviderAuthState> AuthStateAsync() =>
        Task.FromResult(_tokens.Get(Id) is not null ? ProviderAuthState.SignedIn : ProviderAuthState.SignedOut);

    public Task<SignInFlow> StartSignInAsync()
    {
        var pkce = Pkce.Generate();
        _openBrowser(_oauth.MakeAuthorizeUrl(pkce));
        Func<string, Task> submit = async pasted =>
        {
            var token = await _oauth.ExchangeAsync(pasted, pkce);
            _tokens.Save(Id, token);
        };
        return Task.FromResult<SignInFlow>(new SignInFlow.NeedsCode(
            "Approve access in your browser, then paste the code it shows (looks like \"abc123#xyz\").", submit));
    }

    public Task SignOutAsync()
    {
        _tokens.Clear(Id);
        return Task.CompletedTask;
    }

    public async Task<ProviderUsage> FetchUsageAsync()
    {
        var token = _tokens.Get(Id) ?? throw ProviderException.NotSignedIn();
        if (token.NeedsRefresh()) token = await RefreshAndStoreAsync(token);
        token = await EnsurePlanAsync(token);
        try
        {
            return await _fetcher.FetchAsync(token.AccessToken, token.PlanType);
        }
        catch (ProviderException ex) when (ex.Kind == ProviderErrorKind.Unauthorized)
        {
            var refreshed = await EnsurePlanAsync(await RefreshAndStoreAsync(token));
            return await _fetcher.FetchAsync(refreshed.AccessToken, refreshed.PlanType);
        }
    }

    /// <summary>The usage endpoint carries no plan, so fetch the profile once and cache it. Best-effort.</summary>
    private async Task<OAuthToken> EnsurePlanAsync(OAuthToken token)
    {
        if (token.PlanType is not null) return token;
        try
        {
            string? raw = await _profile.FetchRawPlanAsync(token.AccessToken);
            if (raw is null) return token;
            var updated = token with { PlanType = raw };
            _tokens.Save(Id, updated);
            return updated;
        }
        catch
        {
            return token;
        }
    }

    private async Task<OAuthToken> RefreshAndStoreAsync(OAuthToken token)
    {
        try
        {
            var refreshed = await _oauth.RefreshAsync(token);
            refreshed = refreshed with { PlanType = refreshed.PlanType ?? token.PlanType };  // keep cached plan
            _tokens.Save(Id, refreshed);
            return refreshed;
        }
        catch
        {
            _tokens.Clear(Id);
            throw ProviderException.Unauthorized();
        }
    }
}
